# 01. Notebook Setup

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path to project folder

projpath = r'/Users/laineyodette/Documents/_proDocs/professional development/Learning/Learning - Data Analytics/CareerFoundry - Become a Data Analyst/Data Immersion Course/A4 - Python Fundamentals for Data Analysts/Submitted Tasks/2024-01 Instacart Basket Analysis'

# 02. File Import and DF Creation

In [3]:
# Import customers.csv file to create dataframe

df_cust = pd.read_csv(os.path.join(projpath, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [40]:
# Import most recent ords_prods data file

ords_prods_agg = pd.read_pickle(os.path.join(projpath, '02 Data', 'Prepared Data', 'ords_prods_aggregation.pkl'))

In [41]:
# Check dimensions of ord_prods df - should have (32404859, 25) for shape

ords_prods_agg.shape

(32404859, 25)

**Confirmed** ords_prods pulled in properly - all rows included.

In [49]:
# Check column headers of ords_prods df for later use in merge and to check data types

ords_prods_agg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   order_id                    int64  
 1   user_id                     int64  
 2   order_number                int64  
 3   orders_day_of_week          int64  
 4   order_hour_of_day           int64  
 5   days_since_prior_order      float64
 6   customers_first_order       int64  
 7   product_id                  int64  
 8   add_to_cart_order           int64  
 9   reordered                   int64  
 10  product_name                object 
 11  aisle_id                    int64  
 12  department_id               int64  
 13  prices                      float64
 14  price_range_loc             object 
 15  busiest_day                 object 
 16  busiest_days                object 
 17  busiest_period_of_day       object 
 18  max_order                   int64  
 19  loyalty_flag       

# 03. Quality / Consistency Checks + Data Wrangling

## Initial Quality Checks

In [4]:
# Check dimensions of new df

df_cust.shape

(206209, 10)

In [8]:
# Check output - view current column headers

df_cust.head(25)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   
20    29864      Randy     Mills    Male    North Dakota   39    1/1/2017   
21    37744      Wanda     Salas  Female        Arkansas   77    1/1/2017   
22    50000     Sandra     Trejo  Female           Maine   64    1/1/2017   
23    55567     Martha     Hurst  Female  North Carolina   67    1/1/2017   
24    57549      Shawn     Moore    Male         Montana   72    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514  
20             1                           married   44364  
21             1                           married  125977  
22             2                           married  115489  
23             2                           married   48645  
24             0                  divorced/widowed  135302

In [10]:
# Check the values in the date_joined column - all seem to be the same? may not be relevant

df_cust['date_joined'].value_counts()

date_joined
9/17/2018     213
2/10/2018     212
4/1/2019      211
9/21/2019     211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: count, Length: 1187, dtype: int64

**Observation** 

date_joined column has different values. Keep column in case relevant for loyalty or time series evaluation.

In [23]:
# Check data types for each column

df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   First Name            194950 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   Number_of_Dependants  206209 non-null  int64 
 8   Family_Status         206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [34]:
# Check for mixed data types

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].map(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

First Name


In [39]:
# Verify data type of 'First Name' column

df_cust['First Name'].dtype

dtype('O')

In [6]:
# Check output stats

df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

**Initial Evaluation and Planned Action** 

- *Columns not needed?* Plan to keep all columns - each may come in handy for analysis. Note that first/last name are not likely needed for analysis. But could come in handy in recommendations, if any would benefit from having the customer's name.

- *Columns to rename?* Plan to rename 'surnam' to 'last name', 'STATE' to 'state', 'n_dependants' to number_of_dependants', 'fam_status' to 'family_status'.

- *Data type changeds needed?* Plan to update 'user_id' to string. Need to figure out why 'First Name' is returning as a mixed data type. It appears to be a string, but comes back in the "weird" check as mixed. Also a note that the 'user_id' data type in the ords_prods_agg df should also therefore be string.

- *Missing values?* Plan to dive into 'First Name' column as the number of non-null values does not match the total number of rows so there may be some missing. Suspect that this may explain the mixed data type.

- *Any ususual values noticed from describe?* No, all values appear to be reasonable.

## Column Name Changes

In [21]:
# Rename the 'Surnam' column to be more intuitive

df_cust.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [17]:
# Rename the 'STATE' column to be more consistent to other headers

df_cust.rename(columns = {'STATE' : 'State'}, inplace = True)

In [18]:
# Rename the 'n_dependants' column to be more intuitive

df_cust.rename(columns = {'n_dependants' : 'Number_of_Dependants'}, inplace = True)

In [19]:
# Rename the 'fam_status' column to be more intuitive

df_cust.rename(columns = {'fam_status' : 'Family_Status'}, inplace = True)

In [22]:
# Re-check output - view updated column headers

df_cust.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number_of_Dependants Family_Status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

**Confirmed** Column name changes complete.

## Data Type Changes

In [25]:
# Change the data type of the user_id column in customer df to string and override as new type

df_cust['user_id'] = df_cust['user_id'].astype('str')

In [26]:
# Check to ensure data type change has occurred

df_cust['user_id'].dtype

dtype('O')

**Confirmed** Data type change complete.

**Note About 'First Name' Field** 

Suspect that there are missing values in this field, leading to the mixed data type. See next section for evaluation and any action.

In [50]:
# Change the data type of the user_id column in ords_prods_agg df to string and override as new type

ords_prods_agg['user_id'] = ords_prods_agg['user_id'].astype('str')

In [51]:
# Check to ensure data type change has occurred

ords_prods_agg['user_id'].dtype

dtype('O')

**Confirmed** Data type change complete. Note - I did this already on this column back in an earlier task, so not sure why it didn't carry through until now.

## Missing Values Check

In [32]:
# Check for missing values in df

df_cust.isnull().sum()

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
State                       0
Age                         0
date_joined                 0
Number_of_Dependants        0
Family_Status               0
income                      0
dtype: int64

**Observation and Response** 

As suspected, there are missing values in the 'First Name' field. It's about 5% of the data. 11,259 out of 206,209.

While this is a shame, the analysis does not require first names, so no action needed. Where this might become an issue is in recommendations. If there are any recommenation that would target specific customers, having a first name may be helpful. However, we still have the user_id to uniquely identify the customer, so it should be okay.

Also, a note that with the NaN values in this column, this explains the mixed data type returned from that check. Since this field will not be evaluated, no action needed to correct the data type. It is string with NaN values.

## Duplicate Values

In [36]:
# Look for full duplicates in customer df

df_cust_dups = df_cust[df_cust.duplicated()]

In [37]:
# Check for results in duplicate df

df_cust_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, Number_of_Dependants, Family_Status, income]
Index: []

**Observations** No full duplicates found. No action required.

## Verify shape after all consistency and quality checks

In [56]:
# Verify shape after all consistency and quality checks

df_cust.shape

(206209, 10)

# 04. Merge Customer DF into Other Merged DF

- Note common column headers between two dfs (df_cust and ords_prods_agg)

- Use 'user_id' for merge key

In [44]:
# Remove current '_merge' column from ords_prods_agg df so that can check the new merge

ords_prods_agg = ords_prods_agg.drop(columns = ['_merge'])

In [45]:
# Verify column headers of ords_prods df to confirm _merge is removed

for col in ords_prods_agg.columns:
    print(col)

order_id
user_id
order_number
orders_day_of_week
order_hour_of_day
days_since_prior_order
customers_first_order
product_id
add_to_cart_order
reordered
product_name
aisle_id
department_id
prices
price_range_loc
busiest_day
busiest_days
busiest_period_of_day
max_order
loyalty_flag
avg_prod_price
spending_flag
median_days_since_purchase
frequency_flag


In [52]:
# Merge ords_prods_agg with df_cust using 'user_id' as key and include a merge column to indicate which df the row came from

df_merged = ords_prods_agg.merge(df_cust, on = 'user_id', indicator = True)

In [53]:
# Check count of merge flag to verify data came from both dfs

df_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

**Confirmed** Data came from both dfs. Same number of rows as ords_prods_agg base df.

In [54]:
# Check merged df column headers

df_merged.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_prior_order  customers_first_order  product_id  \
0                     NaN                      1         196   
1                    15.0                      0         196   
2                    21.0                      0         196   
3                    29.0                      0         196   
4                    28.0                      0         196   

   add_to_cart_order  reordered  ... First Name  Last Name  Gender    State  \
0                  1          0  ...      Linda     Nguyen  Female  Alabama   
1                  1          1  ...      Linda     Nguyen  Female  Alabama   
2                  1          1  ...      Linda     Nguyen  Female  Alabama   
3                  1          1  ...      Linda     Nguyen  Female  Alabama   
4                  1          1  ...      Linda     Nguyen  Female  Alabama   

  Age date_joined Number_of_Dependants Family_Status  income _merge  
0  31   2/17/2019                    3       married   40423   both  
1  31   2/17/2019                    3       married   40423   both  
2  31   2/17/2019                    3       married   40423   both  
3  31   2/17/2019                    3       married   40423   both  
4  31   2/17/2019                    3       married   40423   both  

[5 rows x 34 columns]

In [55]:
# Check merged df shape

df_merged.shape

(32404859, 34)

# 07. Export merged dataframe

In [57]:
# Export new merged dataframe as pickle file to Prepared Data folder

df_merged.to_pickle(os.path.join(projpath, '02 Data', 'Prepared Data', 'orders_products_all.pkl'))